data/lua/stun.lua

```diff
@@ -0,0 +1,72 @@
+
+local _ = wesnoth.textdomain "wesnoth-units"
+local on_event = wesnoth.require("on_event")
+local old_unit_status = wesnoth.interface.game_display.unit_status
+
+function wesnoth.interface.game_display.unit_status()
+    local u = wesnoth.interface.get_displayed_unit()
+    if not u then return {} end
+    local s = old_unit_status()
+
+    if u.status.stunned then
+        table.insert(s, wml.tag.element{
+            image = "misc/stunned-status-icon.png",
+            tooltip = _"stunned: This unit is stunned. It cannot enforce its Zone of Control."
+        })
+    end
+
+    return s
+end
+
+local function on_hit(weapon, opponent)
+    if not weapon:matches{special_id = 'stun'} then return end
+    if not opponent.zoc then return end
+    local text
+    if opponent.gender == "female" then
+        text = _ "stunned"
+    else
+        text = _ "female^stunned"
+    end
+	local color = stringx.join(',', {'196', '196', '128'})
+    if not wesnoth.interface.is_skipping_messages() then
+        wesnoth.interface.float_label(opponent.x, opponent.y, text, color)
+    end
+    opponent:add_modification('object', {
+        duration = 'turn end',
+        wml.tag.effect{
+            apply_to = 'image_mod',
+            replace = 'CS(50,50,0)'
+        },
+        wml.tag.effect{
+            apply_to = 'zoc',
+            value = false
+        },
+        wml.tag.effect{
+            apply_to = 'status',
+            add = 'stunned'
+        }
+    })
+end
+
+local function attacker_hits(ctx)
+    local weapon = wesnoth.units.create_weapon(wml.get_child(ctx, 'weapon'))
+    local opponent = wesnoth.units.get(ctx.x2, ctx.y2)
+    on_hit(weapon, opponent)
+end
+
+local function defender_hits(ctx)
+    local weapon = wesnoth.units.create_weapon(wml.get_child(ctx, 'second_weapon'))
+    local opponent = wesnoth.units.get(ctx.x1, ctx.y1)
+    on_hit(weapon, opponent)
+end
+
+on_event("attacker_hits", attacker_hits)
+on_event("defender_hits", defender_hits)
+
+-- Removing a status via object doesn't work apparently?
```

Comment:
Does the ZoC get reset at the right time?

Reply:
It certainly looks like it. I saw the tint disappear and the ellipse return to normal at the end of that unit's side's turn.